jose/src/main/java/org/springframework/security/oauth2/jose/JoseHeader.java

```diff
@@ -0,0 +1,183 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.jose;
+
+import org.springframework.security.oauth2.jose.jws.JwsAlgorithm;
+import org.springframework.util.Assert;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.ALG;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.CRIT;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.CTY;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.JKU;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.JWK;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.KID;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.TYP;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.X5C;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.X5T;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.X5T_S256;
+import static org.springframework.security.oauth2.jose.JoseHeaderNames.X5U;
+
+/**
+ *
+ * @author Anoop Garlapati
+ * @author Joe Grandja
+ * @since 0.0.1
+ */
+public final class JoseHeader {
+	private final Map<String, Object> headers;
+
+	private JoseHeader(Map<String, Object> headers) {
+		this.headers = Collections.unmodifiableMap(new LinkedHashMap<>(headers));
+	}
+
+	public JwsAlgorithm getJwsAlgorithm() {
+		return getHeader(ALG);
+	}
+
+	public String getJwkSetUri() {
+		return getHeader(JKU);
+	}
+
+	public Map<String, Object> getJwk() {
+		return getHeader(JWK);
+	}
+
+	public String getKeyId() {
+		return getHeader(KID);
+	}
+
+	public String getX509Uri() {
+		return getHeader(X5U);
+	}
+
+	public List<String> getX509CertificateChain() {
+		return getHeader(X5C);
+	}
+
+	public String getX509SHA1Thumbprint() {
+		return getHeader(X5T);
+	}
+
+	public String getX509SHA256Thumbprint() {
+		return getHeader(X5T_S256);
+	}
+
+	public List<String> getCritical() {
+		return getHeader(CRIT);
+	}
+
+	public String getType() {
+		return getHeader(TYP);
+	}
+
+	public String getContentType() {
+		return getHeader(CTY);
+	}
+
+	public Map<String, Object> getHeaders() {
+		return this.headers;
+	}
+
+	@SuppressWarnings("unchecked")
+	public <T> T getHeader(String name) {
```

Comment:
If this is public, I'd prefer returning an `Object` and having the caller cast it explicitly. Perhaps have a separate private one that auto-casts to make the specific getters neater.

Reply:
I think this is cleaner than casting:

`JwsAlgorithm jwsAlgorithm = joseHeader.getHeader(JoseHeaderNames.ALG)`

vs.

`JwsAlgorithm jwsAlgorithm = (JwsAlgorithm) joseHeader.getHeader(JoseHeaderNames.ALG)`

The caller will know the type.